### PR TITLE
[ENG-1694]: Fix storage memory

### PR DESCRIPTION
**Pull request by ollie-bell**
### Description
Drop masked coefficients via `.where(..., drop=True)` in linear expressions for `RateOfStorageCharge` and `RateOfStorageDischarge`. This significantly reduces memory consumption for `StorageLevel` computation which uses the `LinearExpression.cumsum` method.

The plot below shows an approximately 90% reduction in memory consumption for one model as a result of this change

image

Additionally add in `drop=False` to other linear expressions masking (which is just verbose for the default behaviour) as a placeholder for future potential memory optimisations to be applied. These changes will require further refactoring to deal with misaligned dimensions/coordinates after dropping masked coefficients.

Lastly make tests explicitly use `"highs"` solver.

#### NOTE:

With `drop=False` the memory consumption of the `StorageLevel` expressions scales as:

`O(TECHNOLOGIES * OPERATING_MODES * TIMESLICES)`

whereas with `drop=True` this effectively becomes:

`O(STORAGE_TECHNOLOGIES * STORAGE_OPERATING_MODES * TIMESLICES)`

therefore models with a relatively large number of storage technologies / operating modes will still suffer from high memory consumption. To avoid this entirely, the call to `LinearExpression.cumsum` should be replaced with something else.

### Checklist
- [x] Dependencies install correctly in a clean environment and code executes;
- [ ] Test coverage extended; created tests fail without the change (if possible);
- [x] All tests passing;
- [x] Commits follow a type convention (e.g. https://gist.github.com/brianclements/841ea7bffdb01346392c#type);
- [ ] Extended the README, documentation and/or docstrings, if necessary;

### CODE REVIEW

**Comment by Tomkourou** [comment]
Can you explain what mask dropping does in this context.

**Comment by ollie-bell** [comment]
> Can you explain what mask dropping does in this context.

When you apply a mask the underlying data just get's filled with a bunch of `nan` values - i.e. you have terms in the constraints which are not active, but they still have a memory footprint. In the case of the storage `cumsum` saga this was very significant, and it is still significant elsewhere. With `drop=True` those `nan` values are dropped, along with the memory footprint, and the constraint equations remain identical.

There is still work to be done now however in handling misaligned/missing coordinates which may have disappeared after `drop=True`.

File: tz/osemosys/model/constraints/re_targets.py [hunk 1]
```diff
@@ -53,12 +53,7 @@ def add_re_targets_constraints(ds: xr.Dataset, m: Model, lex: Dict[str, LinearEx
         UseByTechnologyAnnual[r,t,f,y];
     ```
     """
-
-    con = (
-        lex["ProductionAnnualRE"]
-        >= lex["ProductionAnnual"].assign_coords({"FUEL": ds["RETagFuel"] == 1})
-        * ds["REMinProductionTarget"]
-    )
```
Comment (ollie-bell) [inline tz/osemosys/model/constraints/re_targets.py:61]: This change isn't related to masking but I noticed that this `.assign_coords` didn't seem to be doing anything.

File: tz/osemosys/model/constraints/re_targets.py
Comment (ollie-bell) [inline tz/osemosys/model/constraints/re_targets.py:1]: This change isn't related to masking but I noticed that this .assign_coords didn't seem to be doing anything, so I removed it.

File: tz/osemosys/model/constraints/re_targets.py (hunk 1 above)
Comment (ollie-bell) [inline tz/osemosys/model/constraints/re_targets.py:61]: This change isn't related to masking but I noticed that this .assign_coords didn't seem to be doing anything, so I removed it.

**Comment by ollie-bell** [comment]
Decided to revert everything back to `drop=False` except for the exceptional memory consumption resulting from `RateOfStorageCharge` and `RateOfStorageDischarge`. Having added in `drop=False` is just verbose for the default behaviour.

We can revisit reducing memory consumption of other expressions/constraints by turning on `drop=True` in the near future.

File: tz/osemosys/model/linear_expressions/storage.py [hunk 2]
```diff
@@ -5,14 +5,13 @@
...
 def add_lex_storage(ds: xr.Dataset, m: Model, lex: Dict[str, LinearExpression]):
-
     DiscountFactorStorage = (1 + ds["DiscountRateStorage"]) ** (
         1 + ds.coords["YEAR"][-1] - ds.coords["YEAR"][0]
     )
 
     RateOfStorageCharge = (
         (ds["TechnologyToStorage"] * m["RateOfActivity"]).where(
-            (ds["TechnologyToStorage"].notnull()) & (ds["TechnologyToStorage"] != 0)
+            (ds["TechnologyToStorage"].notnull()) & (ds["TechnologyToStorage"] != 0), drop=True
```
Comment (ollie-bell) [inline tz/osemosys/model/linear_expressions/storage.py:14]: This is the actual behaviour change introduced in this PR using `drop=True`

File: tz/osemosys/model/linear_expressions/storage.py [hunk 3]
```diff
@@ -29,13 +28,14 @@ def add_lex_storage(ds: xr.Dataset, m: Model, lex: Dict[str, LinearExpression]):
...
-            & (ds["Conversionls"] != 0)
+            & (ds["Conversionls"] != 0),
+            drop=False,
         )
     ).sum(["TECHNOLOGY", "MODE_OF_OPERATION", "TIMESLICE"])
 
     RateOfStorageDischarge = (
         (ds["TechnologyFromStorage"] * m["RateOfActivity"]).where(
-            (ds["TechnologyFromStorage"].notnull()) & (ds["TechnologyFromStorage"] != 0)
+            (ds["TechnologyFromStorage"].notnull()) & (ds["TechnologyFromStorage"] != 0), drop=True
```
Comment (ollie-bell) [inline tz/osemosys/model/linear_expressions/storage.py:38]: This is the actual behaviour change introduced in this PR using `drop=True`

**Review by Tomkourou — APPROVED** [review]

**Pull request merged**